openmmtools/multistate/multistatesampler.py

```diff
@@ -1195,9 +1195,7 @@ def _propagate_replicas(self):
         # Update all sampler states. For non-0 nodes, this will update only the
         # sampler states associated to the replicas propagated by this node.
         for replica_id, propagated_state in zip(replica_ids, propagated_states):
-            propagated_positions, propagated_box_vectors = propagated_state  # Unpack.
-            self._sampler_states[replica_id].positions = propagated_positions
-            self._sampler_states[replica_id].box_vectors = propagated_box_vectors
+            self._sampler_states[replica_id].__setstate__(propagated_state, ignore_velocities=True)
```

Comment:
Why do you need to use `__setstate__` here?

Reply:
It would be important to document a change in internal data structures like this, since we can confuse `state` as in `SamplerState` as well as serialized state!

Reply:
There's no change in internal data structure. This just serializes the entire sampler state instead of just positions and box vectors (i.e. it serializes also the information on the collective variables so that they are transmitted to MPI node 0).

Reply:
Why can't we just do 
```python
self._sampler_states[replica_id] = propagated_state
```
?

Reply:
We can! But the `__get/setstate()__` methods have a `ignore_velocities` argument which allows us to save bandwidth and MPI would just call those methods anyway to serialize them before sending them over the network.

Reply:
I'd prefer the cleaner code option---if it happens under the hood automatically, we don't need to do it!

Also, we don't always randomize the velocities each iteration, so we only should be ignoring velocities _if_ we are going to be randomizing the velocities each iteration. Otherwise, we're introducing a significant error.

Reply:
To be clear, we don't send velocities only because we don't store velocities on disk. The `SamplerState` in the MPI process remains with velocities whether or not they are randomized.

The `__get/setstate__` methods are used to serialize things throughout the codebase, and they are effectively part of the `States`' API so they don't look weird to me. Besides being more efficient, I also think the code is more transparent about what is really happening here (i.e. we're serializing things and sending them, not just returning them), but I'm happy to change it if we think the loss in performance will be bearable.

Reply:
Just to make sure you caught it, I need `__setstate__` there because [this function](https://github.com/choderalab/openmmtools/blob/fix-mpi-mixing/openmmtools/multistate/multistatesampler.py#L1233) returns the `__getstate__`.